solvebio/cli/data.py

```diff
@@ -735,6 +795,70 @@ def _download_recursive(
                 os.rmdir(local_abs_path)
 
 
+def _resolve_shortcuts_and_get_files(full_path, visited_folders, download_path=None, follow_shortcuts=False):
+    full_path, parts = Object.validate_full_path(full_path)
+    if not download_path:
+        download_path = full_path
+    if not download_path.endswith('/'):
+        download_path += '/'
+
+    results = GlobalSearch().filter(path__prefix=full_path)
+
+    if full_path.endswith("/*"):
+        print(
+            "Folder names ending with '/*' are not supported with "
+            "the --recursive flag, try again with only the folder name."
+        )
+        return set()
+
+    if not follow_shortcuts:
+        return [r for r in results if r.class_name() != "Vault" and not r.is_shortcut]
+
+    files = []
+    for x in results:
+        x.path = download_path + _get_relative_download_path(base_path=full_path,
+                                                             path_to_object=x.full_path,
+                                                             filename=x.name)
+        files.append(x)
+
+    shortcuts = [o for o in files if o.class_name() != "Vault" and o.is_shortcut]
+
+    resolved_shortcuts = []
+    for shortcut in shortcuts:
+        # Global search doesn't return target in the response.
+        # Directly accessing target will always return None.
+        # Call Object.retrieve() to get the full object.
+        target_object = Object.retrieve(shortcut.id).get_target()
+        if not target_object:
+            print("Couldn't find target object for shortcut: {}".format(shortcut.full_path))
+            continue
+        if target_object.is_folder:
+            # todo # implement circular shortcut avoiding
```

Comment:
Is this needed since the backend already prevents circular shortcuts?

Reply:
It looks like `visited_folders` isn't used at all?

Reply:
Yes visited folders isn't used.
I tried implementing some circular shortcut detection with it but it didn't work out.
Left it in there to still keep track of the folders, but it can be removed.

This is one of the circular scenarios I thought of:
<img width="624" alt="image" src="https://github.com/solvebio/solvebio-python/assets/122993108/9dfe5ca2-547d-43c9-990f-7ea9050d66ae">

Something like this isn't handled on the backend.
One way I can propose to guard against this is to implement a graph and use graph traversal to detect potential circular shortcuts.

Reply:
I would add this as a future enhancement